### PR TITLE
MINOR: fix client_compatibility_features_test.py

### PR DESCRIPTION
**Pull request by chia7712**
related to https://github.com/apache/kafka/pull/10105

kafka nodes are set to **older version** so resolved script path is linked to **older assembly**. the 0.10.x versions do not have `org.apache.kafka.tools.ClientCompatibilityTest` so the test gets failed.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by chia7712** [comment]
oh, it tests raft also. will update code to fix that.

File: tests/kafkatest/tests/client/client_compatibility_features_test.py [hunk 1]
```diff
@@ -130,4 +134,7 @@ def run_compatibility_test(self, broker_version, metadata_quorum=quorum.zk):
         self.kafka.set_version(KafkaVersion(broker_version))
         self.kafka.start()
         features = get_broker_features(broker_version)
+        if not self.zk:
+            # this is supported by zkBroker currently
+            features["describe-acls-supported"] = False
```
Comment (rondagostino) [inline tests/kafkatest/tests/client/client_compatibility_features_test.py]: Nice -- better than disabling the test for Raft-based quorums.  Perhaps add to the comment to indicate that this check/disabling is only necessary due to the fact that we are in early access mode with KIP-500 and we should remove the special casing when that his fully implemented?  Aside from that, LGTM -- thank you for the PR!

**Comment by rondagostino** [comment]
This needs to be cherry-picked to 2.8.

File: tests/kafkatest/tests/client/client_compatibility_features_test.py (hunk 1 above)
Comment (chia7712) [reply tests/kafkatest/tests/client/client_compatibility_features_test.py]: will copy that.

**Comment by chia7712** [comment]
@vvcephei  Do you have free time to take a look at this PR? This issue happens in 2.8 also so it would be nice to be included by 2.8.0 :)

**Review by cmccabe — APPROVED** [review]

**Review by cmccabe — CHANGES_REQUESTED** [review]
Wait, can you explain why we need this change now and we didn't need it before?

And if we always want to use the development version of the features test, it seems like this doesn't do that in the KIP-500 case, right?

**Comment by rondagostino** [comment]
I broke this with https://github.com/apache/kafka/pull/10105/files#diff-84e14a0909d232b70f0a957ded161cd077d4dc1d069bbaab8e1bacc8dd2e0572L84-R85.  The test used to always use the ZooKeeper node, and I changed it to use the Kafka node not realizing that the two distribution versions would be different.  We want it to always use at least version 2.7, which we can get by using either the ZooKeeper node for the Zookeeper case or the Kafka node for the Raft case.

**Comment by cmccabe** [comment]
@rondagostino : thanks for the explanation.

It's a bit weird to use the devel version of the tool when ZK is enabled, and the old version when it's not.  I think this will lead to some odd issues down the road.

@chia7712 : Can we just use a separate node for the client, unconditionally?  Then that separate node can always have the devel software.

Or, I suppose, we could change the path that the compat tool is invoked via (maybe that's annoying, though...)

**Comment by rondagostino** [comment]
> use the devel version of the tool when ZK is enabled, and the old version when it's not

The code as amended in this PR will use the devel version when ZooKeeper is enabled (which is the way it always was except after I changed/broke it) and then, when Zookeeper is not enabled -- which is a new case -- it will use the Kafka version, which by definition will be >= 2.8.

**Comment by rondagostino** [comment]
> Can we just use a separate node for the client, unconditionally

We don't have a `Service` for it, and we currently invoke it in a node that we allocate for the test.  With no ZooKeeper all we have is Kafka (version >= 2.8, of course).

I could see a situation where we add functionality to `org.apache.kafka.tools.ClientCompatibilityTest` in some future version and then we try to run that against Kafka v2.8 -- and then it'll go BOOM! as you suggest.

File: tests/kafkatest/tests/client/client_compatibility_features_test.py [hunk 2]
```diff
@@ -81,8 +81,12 @@ def __init__(self, test_context):
...
 
     def invoke_compatibility_program(self, features):
-        # Run the compatibility test on the first Kafka node.
-        node = self.kafka.nodes[0]
+        if self.zk:
+            # kafka nodes are set to older version so resolved script path is linked to older assembly.
+            # run the compatibility test on the first zk node to get script path linked to latest(dev) assembly.
+            node = self.zk.nodes[0]
+        else:
+            node = self.kafka.nodes[0]
```
Comment (rondagostino) [inline tests/kafkatest/tests/client/client_compatibility_features_test.py]: We can use the dev version of the tool on the Kafka node via code like this:
```
        node = self.kafka.nodes[0]
        cmd = ("%s org.apache.kafka.tools.ClientCompatibilityTest "
               "--bootstrap-server %s "
               "--num-cluster-nodes %d "
               "--topic %s " % (self.dev_script_path,
                               self.kafka.bootstrap_servers(),
                               len(self.kafka.nodes),
                               list(self.topics.keys())[0]))
```

And then further down we can define the DEV script path like this:

```
        # Always use the latest version of org.apache.kafka.tools.ClientCompatibilityTest
        # so store away the path to the DEV version before we set the Kafka version
        self.dev_script_path = self.kafka.path.script("kafka-run-class.sh", self.kafka.nodes[0])
        self.kafka.set_version(KafkaVersion(broker_version))
```

I tested this locally and it solves the problem.

Comment (chia7712) [reply tests/kafkatest/tests/client/client_compatibility_features_test.py]: copy that

**Comment by chia7712** [comment]
@cmccabe @rondagostino thanks for all your comments (and nice explanation)! I have update code according to reviews. please take a look.

File: tests/kafkatest/tests/client/client_compatibility_features_test.py [hunk 3]
```diff
@@ -79,14 +79,16 @@ def __init__(self, test_context):
             "replication-factor": 3
             }}
         self.kafka = KafkaService(test_context, num_nodes=3, zk=self.zk, topics=self.topics)
+        # Always use the latest version of org.apache.kafka.tools.ClientCompatibilityTest
+        # so store away the path to the DEV version before we set the Kafka version
+        self.dev_script_path = self.kafka.path.script("kafka-run-class.sh", self.kafka.nodes[0])
 
     def invoke_compatibility_program(self, features):
-        # Run the compatibility test on the first Kafka node.
```
Comment (rondagostino) [inline tests/kafkatest/tests/client/client_compatibility_features_test.py:84]: nit: no need to remove this comment.  Otherwise LGTM -- thank you again!

Comment (chia7712) [reply tests/kafkatest/tests/client/client_compatibility_features_test.py:84]: will copy that.

**Comment by cmccabe** [comment]
Thanks, @chia7712 !

**Review by cmccabe — APPROVED** [review]

**Comment by chia7712** [comment]
push to trunk and 2.8

**Pull request merged**